quantumblacklabs/kedro-viz
# [KED-2328] Add code panel

## Description

Adds a code panel to the metadata sidebar.

<img width="1292" alt="code sidebar screenshot" src="https://user-images.githubusercontent.com/45365769/103925846-07602480-5110-11eb-8955-24137c363da1.png">

## Development notes

- Adds the code panel component
- Uses highlight.js for code highlighting
- Handles both light and dark highlighting themes
- Improves the responsive approach so that the chart moves and scales to fit
- Adds a feature flag named `code`

## QA notes

Testing requires a live Kedro project running through the viz API.

The `code` feature flag must be enabled before testing.

When a function node is clicked and code is available from the API, the code panel toggle will appear at the top of the metadata panel. Other node types do not show the toggle.

Clicking the toggle will open the panel and show highlighted code.

Test at various browser sizes and with different sidebar configurations, as the chart will responsively scale to fit the space available until a minimum chart size.

The code panel is intentionally a fixed width, if code is wider than the panel it can be scrolled horizontally.

When the clicked node is changed or removed, the code panel will close itself.

Test using both themes.

## Checklist


## Legal notice


- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
